synapse/storage/engines/sqlite.py

```diff
@@ -81,10 +88,15 @@ def on_new_connection(self, db_conn: "LoggingDatabaseConnection") -> None:
         from synapse.storage.prepare_database import prepare_database
 
         if self._is_in_memory:
-            # In memory databases need to be rebuilt each time. Ideally we'd
-            # reuse the same connection as we do when starting up, but that
-            # would involve using adbapi before we have started the reactor.
-            prepare_database(db_conn, self, config=None)
+            if self._prepped_conn is not None:
+                # Initialise the new DB from the pre-prepared DB.
+                assert isinstance(db_conn.conn, sqlite3.Connection)
+                self._prepped_conn.backup(db_conn.conn)
+            else:
+                # In memory databases need to be rebuilt each time. Ideally we'd
+                # reuse the same connection as we do when starting up, but that
+                # would involve using adbapi before we have started the reactor.
```

Comment:
I keep stumbling over this comment for some reason. Is it still fully true after this change? It looks like it was just indented.

Reply:
Hmm, maybe it should be moved up a layer? We do need to rebuild the DB each time, its just one method of doing that is copying from the template?

Reply:
So `prepped_conn` is a connection to an in-memory database that we've previously prepared via `prepare_database`. If we have one of those, we can just copy that DB into the new DB (via `.backup(..)`), otherwise we need to call `prepare_database` on the new connection.  Either way, the new DB/connection ends up having been prepared, its just in one case we've used an existing "template" DB (which is what they're called in postgres land), and in the other we manually prepare it.

I've updated the comment a bit, not sure if its easier to understand?

Reply:
Yes, that helps a bit. 👍 

I do find it a bit confusing that we're using postgres terms for in-memory databases which are sqlite only, I think? 😄

Reply:
Hmm, I thought thinking of the pre-prepared DB as a "template" for new DBs seemed like a good term :shrug:

Reply:
Yeah that's fine. I think the comment is probably clear enough now FTR